zerver/lib/reminders.py

```diff
@@ -0,0 +1,66 @@
+from django.conf import settings
+from django.utils.translation import gettext as _
+
+from zerver.actions.message_send import internal_send_private_message
+from zerver.lib.exceptions import JsonableError
+from zerver.lib.markdown.fenced_code import get_unused_fence
+from zerver.lib.mention import silent_mention_syntax_for_user
+from zerver.lib.message import access_message
+from zerver.lib.message_cache import MessageDict
+from zerver.lib.url_encoding import near_message_url, topic_narrow_url
+from zerver.models import Message, Stream, UserProfile
+from zerver.models.users import get_system_bot
+
+
+def get_reminder_formatted_content(
+    message: Message, current_user: UserProfile, stream: Stream | None
+) -> str:
+    if message.is_stream_message():
+        assert stream is not None
+        narrow_link = topic_narrow_url(
+            realm=current_user.realm,
+            stream=stream,
+            topic_name=message.topic_name(),
+        )
+        content = _(
+            "You requested a reminder for the following message sent to [{stream_name} > {topic_name}]({narrow_link})."
+        ).format(
+            stream_name=stream.name,
+            topic_name=message.topic_name(),
+            narrow_link=narrow_link,
+        )
+    else:
+        content = _("You requested a reminder for the following private message.")
+
+    # Format the message content as a quote.
+    content += "\n\n"
+    content += _("{user_silent_mention} [said]({conversation_url}):").format(
+        user_silent_mention=silent_mention_syntax_for_user(message.sender),
+        conversation_url=near_message_url(current_user.realm, MessageDict.wide_dict(message)),
+    )
+    content += "\n"
+    fence = get_unused_fence(content)
+    quoted_message = f"{fence}quote\n{message.content}\n{fence}"
+    content += quoted_message
+    return content
+
+
+def access_message_for_reminder(current_user: UserProfile, message_id: int) -> Message:
+    try:
+        message = access_message(current_user, message_id, is_modifying_message=False)
+    except JsonableError as error:
+        internal_send_private_message(
+            get_system_bot(settings.NOTIFICATION_BOT, current_user.realm.id),
+            current_user,
+            _(
+                "You asked for a reminder about a [message with ID {message_id}]({message_link}), but either the message has been deleted or you no longer have access to the message."
+            ).format(
```

Comment:
Thinking more about the error experience, I think it might be better to just send the original reminder text if the `access_message` check fails. Reasoning is:
- Security/UX wise, it'd work exactly like if you did quoted a message and put that into a scheduled message.
- It seems a lot simpler to not have this weird error case.

I think probably other than that, we don't particularly expect to have errors when trying to deliver the message -- the only ways Notification Bot will be unable to send you a DM are if your account is deleted, and then (1) probably the reminder should have been removed during account deactivation and (2) there's nobody to deliver it to.

Reply:
Okay, we send the stored content if the original message is inaccessible.

Reply:
Yeah, done.